Virts 501 validate reverse conns

unkempthenry opened the pull request:
Prevents terminal plugin from believing any connection request (from scans / etc) is an incoming shell.  The reverse payload now sends a sha256 of its 'key' variable.  Previously, this value was only used to change the hash of the file however. 

Changes:
* `reverse.go` now sends a hash of its `key` value when it first connects -- the terminal plugin now only accepts shells that send a hash of a known key value. 
* Reproduced the on-demand compilation logic from the sandcat plugin, so that works here again now and updated the shell spawning ability to work with this logic.
* terminal will add a new `terminal_key` table to `core.db` which is used to track previously used keys. 

NOTE1: The downside of doing it like is that dynamically compiled reverse payloads can't be shared across caldera servers, or across server restarts with a non-persistent database.  However, the pre-compiled binaries use a default key value, which the server will still accept (at least for now), so those can still be shared around.

NOTE2: This doesn't offer reliable authentication between the reverse payload and server, but it lays the groundwork for this; in a future change, the payload and caldera server could exchange nonces and use them plus the shared key value to do a mutually authenticating handshake safe from replay protection.

Commit:
fix terminal after human-in-the-loop changes

Commit:
dynamically compile reverse.go and validate that it has a hash of its key value

Commit:
save go compile keys in db

Commit:
add sql to create table

wbooth commented on app/terminal/shell.py:
```diff
@@ -35,8 +35,10 @@ def __init__(self, services):
-            except Exception:
-                self.console.line('Bad command', 'red')
+            except Exception as e:
+                import traceback
```
imports should be at the top of the file

wbooth commented:
some changes went in on the day this PR was submitted, can you fix:
see https://github.com/mitre/terminal/pull/7/files

Commit:
Update to be compatible with recent changes

Commit:
Remove use of sqlite table

unkempthenry commented:
Updated to work with recent changes.  Also, removed the use of a sqlite table to store keys used to generate the payloads and stores them in memory instead.  This means that if the caldera server is restarted it will refuse connections from any previously dynamically compiled payloads, which makes me less sure about this approach.

wbooth commented:
Discussed, separating into multiple prs
